signal/micro/kernels/stacker.cc

```diff
@@ -0,0 +1,171 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <stdint.h>
+
+#include "signal/src/circular_buffer.h"
+#include "tensorflow/lite/kernels/internal/tensor_ctypes.h"
+#include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/micro/flatbuffer_utils.h"
+#include "tensorflow/lite/micro/kernels/kernel_util.h"
+#include "tensorflow/lite/micro/memory_helpers.h"
+#include "tensorflow/lite/micro/micro_utils.h"
+
+namespace tflite {
+namespace {
+
+constexpr int kInputTensor = 0;
+constexpr int kOutputTensor = 0;
+constexpr int kOutputValidTensor = 1;
+
+// Indices into the init flexbuffer's vector.
+// The parameter's name is in the comment that follows.
+// Elements in the vectors are ordered alphabetically by parameter name.
+constexpr int kNumChannelsIndex = 0;          // 'num_channels'
+constexpr int kStackerLeftContextIndex = 1;   // 'stacker_left_context'
+constexpr int kStackerRightContextIndex = 2;  // 'stacker_right_context'
+constexpr int kStackerStepIndex = 3;          // 'stacker_step'
+
+struct TFLMSignalStackerParams {
+  int32_t num_channels;
+  int32_t stacker_left_context;
+  int32_t stacker_right_context;
+  int32_t stacker_step;
+
+  size_t buffer_size;
+  size_t step_size;
+  bool stacker_has_first_frame;
+
+  int8_t* state;
+  tflm_signal::CircularBuffer* circular_buffer;
+};
+
+void* Init(TfLiteContext* context, const char* buffer, size_t length) {
+  const uint8_t* buffer_t = reinterpret_cast<const uint8_t*>(buffer);
+
+  auto* params =
+      static_cast<TFLMSignalStackerParams*>(context->AllocatePersistentBuffer(
+          context, sizeof(TFLMSignalStackerParams)));
+  if (params == nullptr) {
+    return params;
+  }
+
+  tflite::FlexbufferWrapper fbw(buffer_t, length);
+  params->num_channels = fbw.ElementAsInt32(kNumChannelsIndex);
+  params->stacker_left_context = fbw.ElementAsInt32(kStackerLeftContextIndex);
+  params->stacker_right_context = fbw.ElementAsInt32(kStackerRightContextIndex);
+  params->stacker_step = fbw.ElementAsInt32(kStackerStepIndex);
+
+  params->buffer_size =
+      params->num_channels *
+      (params->stacker_left_context + params->stacker_right_context + 1);
+  params->step_size = params->num_channels * params->stacker_step;
+  params->stacker_has_first_frame = false;
+
+  size_t state_size =
+      tflm_signal::CircularBufferGetNeededMemory(params->buffer_size);
+  params->state = static_cast<int8_t*>(
+      context->AllocatePersistentBuffer(context, sizeof(int8_t) * state_size));
```

Comment:
Can we add a nullptr check here and return nullptr?

Reply:
done !